llext: export DT devices to extensions

pillo79 opened the pull request:
This PR adds support for automatic export of all Zephyr devices to extensions via a new Kconfig option, `CONFIG_LLEXT_EXPORT_DEVICES`. When this option is enabled, all Devicetree macros in `zephyr/device.h` can be used in extensions as well to inspect and interact with the devices configured in the current build.

Calling `EXPORT_SYMBOL` with anything more complicated than a simple identifier resulted in macro expansion issues. This has been fixed by adding a level of indirection, moving the actual symbol definition code to private `Z_` macros. A number of related comments have been updated. 

New test code has been added to properly verify the new support. The test retrieves the device object associated with the Zephyr console via `DT_CHOSEN` macros and verifies it matches with the one obtained at runtime by calling `device_get_binding()`. This check is currently unsupported on Xtensa since the `qemu_xtensa` platform defines no usable DT devices.    

Also fixes 2 minor issues in `device.h`:
- The error message for `Z_DEVICE_NAME_CHECK` included a call to `DEVICE_NAME_GET`, but the argument was already a string. This resulted in a confusing error message being displayed when the check failed.
- There was a misleading comment that implied the name returned by `Z_DEVICE_DT_DEV_ID` was used in `device_get_binding()` instead of the DT path, and must thus be limited to `Z_DEVICE_MAX_NAME_LEN` chars. 
This is not the case (anymore?): `device_get_binding()` matches neither paths nor identifiers, but [either DT _labels_ or DT _full names_](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/device.h#L152-L153) as returned by the `DEVICE_DT_NAME` macro [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/device.h#L199).

pillo79 committed:
device.h: fix bad error message for Z_DEVICE_NAME_CHECK

The error message for Z_DEVICE_NAME_CHECK was needlessly calling
DEVICE_NAME_GET() on the "name" string, resulting in a confusing message
being displayed.

Remove the DEVICE_NAME_GET call so that the error message shown by the
compiler contains the actual name.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: export DT devices to extensions

This change adds a new configuration option, LLEXT_EXPORT_DEVICES, which
enables exporting all devices defined in the device tree to llext
extensions. When enabled, all devices are made available to extensions
via the standard DT_ / DEVICE_* macros.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
device.h: fix misleading comment

The comment for Z_DEVICE_DT_DEV_ID() is misleading: device_get_binding()
compares its inputs with values generated by the DEVICE_DT_NAME() macro,
which expands to the node's DT label or to the node's full name.

This patch updates the comment to reflect the actual behavior of
Z_DEVICE_DT_DEV_ID().

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: test exporting DT devices to extensions

Add a new test to the simple llext test suite that verifies that DT
devices are properly exported to extensions. This is done by obtaining a
device handle from the DT and then trying to get the same handle at
runtime using the device_get_binding() API.

This test is optional because the current qemu_xtensa platform does not
have any usable DT devices to test with.

The LLEXT heap size is also increased to 64 kbytes to avoid heap
exhaustion when running the new test on mps2_an385.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext: add wrapper to EXPORT_SYMBOL / LL_EXTENSION_SYMBOL

This patch adds an indirection level moving the current code from inside
the EXPORT_SYMBOL and LL_EXTENSION_SYMBOL macros to private Z_* macros.
This is done to allow for the official APIs to properly expand more
complex arguments, such as the DT function-like macros.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

lyakh commented on include/zephyr/device.h:
```diff
@@ -91,10 +91,8 @@ typedef int16_t device_handle_t;
- * dev_id from a devicetree node while staying within this maximum
- * size.
+/* This macro synthesizes a unique dev_id from a devicetree node by using
+ * the node's dependency ordinal.
```
I think you're modifying a comment for `Z_DEVICE_DT_DEV_ID()` and not for `DEVICE_NAME_GET()` as the commit description seems to suggest?

pillo79 replied:
Thanks for spotting, will fix the commit description along with the reviews :+1:

fabiobaltieri commented:
@gmarull can you take a look?

gmarull commented on include/zephyr/device.h:
```diff
@@ -999,7 +999,7 @@ device_get_dt_nodelabels(const struct device *dev)
 #define Z_DEVICE_NAME_CHECK(name)                                              \
 	BUILD_ASSERT(sizeof(Z_STRINGIFY(name)) <= Z_DEVICE_MAX_NAME_LEN,       \
-			    Z_STRINGIFY(DEVICE_NAME_GET(name)) " too long")
+			    Z_STRINGIFY(name) " too long")
```
ack

gmarull commented on include/zephyr/device.h:
```diff
@@ -197,7 +207,8 @@ typedef int16_t device_handle_t;
 			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)),     \
-			__VA_ARGS__)
+			__VA_ARGS__);                                          \
+	Z_DEVICE_EXPORT(node_id)
```
you can use `IF_ENABLED(CONFIG_LLEXT_EXPORT_DEVICES, (EXPORT_SYMBOL(DEVICE_DT_NAME_GET(node_id))))`

pillo79 replied:
I could, but the separate macro is easier to extend in the (short) future; I have a draft of a more complex proposal (https://github.com/zephyrproject-rtos/zephyr/pull/77799) where there are two different implementations for `Z_DEVICE_EXPORT`. 
For this reason I'm proposing to keep the separate macro, but included within an `IF_ENABLED` in the parent.

gmarull commented on include/zephyr/device.h:
```diff
@@ -12,6 +12,7 @@
 #include <zephyr/devicetree.h>
 #include <zephyr/init.h>
 #include <zephyr/linker/sections.h>
+#include <zephyr/llext/symbol.h>
```
conditional include?

gmarull commented on include/zephyr/device.h:
```diff
@@ -99,6 +100,14 @@ typedef int16_t device_handle_t;
  */
 #define Z_DEVICE_DT_DEV_ID(node_id) _CONCAT(dts_ord_, DT_DEP_ORD(node_id))
 
+#if defined(CONFIG_LLEXT_EXPORT_DEVICES)
+/* Export device identifiers using the builtin name */
+#define Z_DEVICE_EXPORT(node_id) EXPORT_SYMBOL(DEVICE_DT_NAME_GET(node_id))
+#else
+/* Do not export device identifiers */
+#define Z_DEVICE_EXPORT(node_id)
+#endif
+
```
see my other comment

gmarull commented on include/zephyr/device.h:
```diff
@@ -171,7 +180,8 @@ typedef int16_t device_handle_t;
  * The device is declared with extern visibility, so a pointer to a global
  * device object can be obtained with `DEVICE_DT_GET(node_id)` from any source
- * file that includes `<zephyr/device.h>`. Before using the pointer, the
+ * file that includes `<zephyr/device.h>` (even from extensions, when @c
+ * CONFIG_LLEXT_EXPORT_DEVICES is enabled). Before using the pointer, the
```
wrong doxygen use, use `@kconfig{CONFIG_...}`

gmarull commented on include/zephyr/device.h:
```diff
@@ -197,7 +207,8 @@ typedef int16_t device_handle_t;
 			level, prio, api,                                      \
 			&Z_DEVICE_STATE_NAME(Z_DEVICE_DT_DEV_ID(node_id)),     \
-			__VA_ARGS__)
+			__VA_ARGS__);                                          \
```
shouldn't trailing `;` be conditioned to the next statement? if not, move export above.

pillo79 commented:
**v2**:
- new proposal for  `Z_DEVICE_EXPORT`
- implemented all other notes by @gmarull 
- renamed new private macros from `macro_IMPL` to `Z_macro` to keep in line with Zephyr internals

gmarull reviewed (approved):
ack device changes
